README.md
[![Go Report Card](https://goreportcard.com/badge/github.com/postfinance/httpclient)](https://goreportcard.com/report/github.com/postfinance/httpclient)
[![GoDoc](https://godoc.org/github.com/postfinance/httpclient?status.svg)](https://godoc.org/github.com/postfinance/httpclient)
[![Build Status](https://github.com/postfinance/httpclient/workflows/build/badge.svg)](https://github.com/postfinance/httpclient/actions)
[![Coverage Status](https://coveralls.io/repos/github/postfinance/httpclient/badge.svg?branch=master)](https://coveralls.io/github/postfinance/httpclient?branch=master)

# httpclient

Generates a HTTP client from a service definition (interface). The created client is ready to use in production with many configuration options and sensible defaults.

## Requirements

Go 1.7

## Installation

    # go get -u github.com/postfinance/httpclient/cmd/httpclient-gen-go

## Usage
See [example](example/)


CHANGELOG.md
## 0.1.6 (2022-07-26)


### Bug Fixes

* **common**: clone header to prevent from endless appending if header is set (fe06d4baf84ce157882be43a2bf385895837a90b)





httpclient_test.go
package httpclient

import (
	"bufio"
	"bytes"
	"context"
	"io"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"net/http/httputil"
	"net/url"
	"testing"

	"golang.org/x/time/rate"

	"github.com/moul/http2curl"
	"github.com/stretchr/testify/assert"
)

type options struct {
	Page    int    `url:"page,omitempty"`
	PerPage int    `url:"per_page,omitempty"`
	Search  string `url:"search,omitempty"`
}

type message struct {
	Text string
}

func (m message) String() string {
	return m.Text
}

// nolint: gochecknoglobals
var (
	baseurl        = "https://hostname.domain"
	baseurlInvalid = "https://show you how deep the rabbit hole goes.domain"
	username       = "cognitive"
	password       = "distortions"
	contentType    = "application/yaml"
	testMessage    = message{
		Text: "it's only rock'n'roll",
	}
)

// nolint: funlen, gocognit, gocyclo
func TestClient(t *testing.T) {
	t.Run("query options", func(t *testing.T) {
		opt := options{1, 10, "name=testHost"}
		u, err := QueryOptions(baseurl, opt)
		assert.Nil(t, err)
		assert.Equal(t, "https://hostname.domain?page=1&per_page=10&search=name%3DtestHost", u)
	})

	t.Run("query options nil", func(t *testing.T) {
		var opt *options
		u, err := QueryOptions(baseurl, opt)
		assert.Nil(t, err)
		assert.Equal(t, baseurl, u)
	})

	t.Run("baseurl invalid", func(t *testing.T) {
		u, err := QueryOptions(baseurlInvalid, nil)
		assert.NotNil(t, err)
		assert.Equal(t, baseurlInvalid, u)
	})

	t.Run("query options not query values", func(t *testing.T) {
		opt := 42
		u, err := QueryOptions(baseurl, opt)
		assert.NotNil(t, err)
		assert.Equal(t, baseurl, u)
	})

	t.Run("new client invalid baseurl", func(t *testing.T) {
		_, err := New(baseurlInvalid)
		assert.NotNil(t, err)
	})

	t.Run("new client valid baseurl", func(t *testing.T) {
		c, err := New(baseurl)
		assert.Nil(t, err)
		assert.NotNil(t, c)
		u, _ := url.Parse(baseurl)
		assert.Equal(t, u, c.BaseURL)
	})

	t.Run("new client valid baseurl invalid username", func(t *testing.T) {
		_, err := New(baseurl, WithUsername(""))
		assert.NotNil(t, err)
	})

	t.Run("new client valid baseurl valid username", func(t *testing.T) {
		c, err := New(baseurl, WithUsername(username))
		assert.Nil(t, err)
		assert.NotNil(t, c)
		assert.Equal(t, username, c.username)
	})

	t.Run("new client valid baseurl invalid password", func(t *testing.T) {
		_, err := New(baseurl, WithPassword(""))
		assert.NotNil(t, err)
	})

	t.Run("new client valid baseurl valid password", func(t *testing.T) {
		c, err := New(baseurl, WithPassword(password))
		assert.Nil(t, err)
		assert.NotNil(t, c)
		assert.Equal(t, password, c.password)
	})

	t.Run("new client with headers", func(t *testing.T) {
		customHeader := http.Header{
			"X-Requested-By": []string{"test"},
		}
		c, err := New(baseurl, WithHeader(customHeader))
		assert.Nil(t, err)
		assert.NotNil(t, c)

		req, err := c.NewRequest(http.MethodGet, "/test", nil)
		assert.Nil(t, err)
		assert.Equal(t, req.Header["X-Requested-By"], []string{"test"})
		assert.Contains(t, req.Header, "Content-Type")
		assert.Contains(t, req.Header, "Accept")
	})

	t.Run("new client with headers and basic auth", func(t *testing.T) {
		username := "user1"  // nolint: govet
		password := "123456" // nolint: govet
		fakeAuthHeader := http.Header{
			"Authorization": []string{"fake"},
		}

		c, err := New(baseurl, WithHeader(fakeAuthHeader), WithUsername(username), WithPassword(password))
		assert.Nil(t, err)
		assert.NotNil(t, c)

		req, err := c.NewRequest(http.MethodGet, "/test", nil)
		assert.Nil(t, err)
		user, passwd, ok := req.BasicAuth()
		assert.True(t, ok)
		assert.Equal(t, user, username)
		assert.Equal(t, passwd, passwd)
	})

	t.Run("new client valid baseurl valid HTTP client", func(t *testing.T) {
		httpC := &http.Client{}
		c, err := New(baseurl, WithHTTPClient(httpC))
		assert.Nil(t, err)
		assert.NotNil(t, c)
		assert.True(t, httpC == c.client)
	})

	t.Run("new client valid baseurl valid rate limiter", func(t *testing.T) {
		limiter := rate.NewLimiter(1.0, 1)
		c, err := New(baseurl, WithRateLimiter(limiter))
		assert.Nil(t, err)
		assert.NotNil(t, c)
		assert.True(t, limiter == c.limiter)
	})

	t.Run("new client valid baseurl valid content type", func(t *testing.T) {
		_, err := New(baseurl, WithContentType(contentType))
		assert.Nil(t, err)
	})

	t.Run("new client valid baseurl invalid content type", func(t *testing.T) {
		_, err := New(baseurl, WithContentType(""))
		assert.NotNil(t, err)
	})

	t.Run("new request with Marshaler == nil", func(t *testing.T) {
		defer func() {
			assert.NotNil(t, recover())
		}()
		c, _ := New(baseurl)
		c.Marshaler = nil
		_, _ = c.NewRequest(http.MethodGet, "node", nil)
		assert.Fail(t, "NewRequest did not panic")
	})

	t.Run("new request with RequestCallback == nil", func(t *testing.T) {
		defer func() {
			assert.NotNil(t, recover())
		}()
		c, _ := New(baseurl)
		c.RequestCallback = nil
		_, _ = c.NewRequest(http.MethodGet, "node", nil)
		assert.Fail(t, "NewRequest did not panic")
	})

	t.Run("new request using RequestCallback to dump request", func(t *testing.T) {
		c, _ := New(baseurl)
		var dump []byte
		c.RequestCallback = func(r *http.Request) *http.Request {
			dump, _ = httputil.DumpRequestOut(r, true)
			return r
		}
		_, _ = c.NewRequest(http.MethodGet, "node", nil)
		assert.True(t, len(dump) > 0)
	})

	t.Run("new request using RequestCallback to dump request", func(t *testing.T) {
		c, _ := New(baseurl)
		var command *http2curl.CurlCommand
		c.RequestCallback = func(r *http.Request) *http.Request {
			command, _ = http2curl.GetCurlCommand(r)
			return r
		}
		_, _ = c.NewRequest(http.MethodGet, "node", nil)
		assert.True(t, len(command.String()) > 0)
	})

	t.Run("new request without basic auth", func(t *testing.T) {
		c, err := New(baseurl)
		assert.Nil(t, err)
		assert.NotNil(t, c)
		req, err := c.NewRequest(http.MethodGet, "node", nil)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		_, _, ok := req.BasicAuth()
		assert.False(t, ok)
	})

	t.Run("new request with basic auth", func(t *testing.T) {
		c, err := New(baseurl, WithUsername(username), WithPassword(password))
		assert.Nil(t, err)
		assert.NotNil(t, c)
		req, err := c.NewRequest(http.MethodGet, "node", nil)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		u, p, ok := req.BasicAuth()
		assert.True(t, ok)
		assert.Equal(t, username, u)
		assert.Equal(t, password, p)
	})

	t.Run("new request with content type application/json", func(t *testing.T) {
		c, err := New(baseurl)
		assert.Nil(t, err)
		assert.NotNil(t, c)
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		buf := new(bytes.Buffer)
		_, err = io.Copy(buf, req.Body)
		assert.Nil(t, err)
		assert.Equal(t, "{\"Text\":\"it's only rock'n'roll\"}\n", buf.String())
	})

	t.Run("new request with content type application/yaml", func(t *testing.T) {
		c, err := New(baseurl)
		c.ContentType = ContentTypeYAML
		assert.Nil(t, err)
		assert.NotNil(t, c)
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		buf := new(bytes.Buffer)
		_, err = io.Copy(buf, req.Body)
		assert.Nil(t, err)
		assert.Equal(t, "text: it's only rock'n'roll\n", buf.String())
	})

	t.Run("new request with content type text/plain", func(t *testing.T) {
		c, err := New(baseurl)
		c.ContentType = ContentTypeText
		assert.Nil(t, err)
		assert.NotNil(t, c)
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		buf := new(bytes.Buffer)
		_, err = io.Copy(buf, req.Body)
		assert.Nil(t, err)
		assert.Equal(t, "it's only rock'n'roll", buf.String())
	})

	t.Run("new request with unknown content type ", func(t *testing.T) {
		c, err := New(baseurl)
		c.ContentType = "unknown/unknown"
		assert.Nil(t, err)
		assert.NotNil(t, c)
		_, err = c.NewRequest(http.MethodGet, "node", struct{ Message string }{Message: "it's only rock'n'roll"})
		assert.NotNil(t, err)
	})

	// Test server
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/node" {
			http.Error(w, "invalid", http.StatusNotFound)
			return
		}
		// check post data
		body, err := ioutil.ReadAll(r.Body)
		if err != nil {
			t.Fatal(err)
		}
		switch r.Header.Get("Content-Type") {
		case ContentTypeJSON:
			w.Header().Set("Content-Type", ContentTypeJSON)
		case ContentTypeYAML:
			w.Header().Set("Content-Type", ContentTypeYAML)
		case ContentTypeText:
			w.Header().Set("Content-Type", ContentTypeText)
		default:
			w.Header().Set("Content-Type", "unknown/unknown")
		}

		_, err = w.Write(body)
		if err != nil {
			t.Fatal(err)
		}
	}))
	defer ts.Close()
	t.Log("test server URL:", ts.URL)

	t.Run("do a request with Unmarshaler == nil", func(t *testing.T) {
		defer func() {
			assert.NotNil(t, recover())
		}()
		c, _ := New(ts.URL)
		c.Unmarshaler = nil
		ctx := context.Background()
		act := &message{}
		req, err := c.NewRequest(http.MethodGet, "node", act)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		resp, _ := c.Do(ctx, req, act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Fail(t, "Do did not panic")
	})

	t.Run("do a request with ResponseCallback == nil", func(t *testing.T) {
		defer func() {
			assert.NotNil(t, recover())
		}()
		c, _ := New(ts.URL)
		c.ResponseCallback = nil
		ctx := context.Background()
		act := &message{}
		req, err := c.NewRequest(http.MethodGet, "node", act)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		resp, _ := c.Do(ctx, req, act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Fail(t, "Do did not panic")
	})

	t.Run("do a request with a writer", func(t *testing.T) {
		c, _ := New(ts.URL)
		ctx := context.Background()
		act := &message{}
		req, err := c.NewRequest(http.MethodGet, "node", act)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		var buf bytes.Buffer
		resp, err := c.Do(ctx, req, bufio.NewWriter(&buf))
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Nil(t, err)
		assert.True(t, buf.Len() > 0)
	})

	t.Run("do a request with content type application/json", func(t *testing.T) {
		c, _ := New(ts.URL)
		ctx := context.Background()
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		act := &message{}
		resp, err := c.Do(ctx, req, act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Nil(t, err)
		assert.Equal(t, &testMessage, act)
	})

	t.Run("do a request with content type application/yaml", func(t *testing.T) {
		c, _ := New(ts.URL)
		c.ContentType = ContentTypeYAML
		ctx := context.Background()
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		act := &message{}
		resp, err := c.Do(ctx, req, act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Nil(t, err)
		assert.Equal(t, &testMessage, act)
	})

	t.Run("do a request with content type text/plain", func(t *testing.T) {
		c, _ := New(ts.URL)
		c.ContentType = ContentTypeText
		ctx := context.Background()
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		act := ""
		resp, err := c.Do(ctx, req, &act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Nil(t, err)
		assert.Equal(t, testMessage.String(), act)
	})

	t.Run("do a request with content type text/plain and provide non-string-type", func(t *testing.T) {
		c, _ := New(ts.URL)
		c.ContentType = ContentTypeText
		ctx := context.Background()
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		act := 42
		resp, err := c.Do(ctx, req, &act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.NotNil(t, err)
	})

	t.Run("do a request with error in response using default ResponseCallback", func(t *testing.T) {
		c, _ := New(ts.URL)
		ctx := context.Background()
		act := &message{}
		req, err := c.NewRequest(http.MethodGet, "invalid", act)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		var buf bytes.Buffer
		resp, err := c.Do(ctx, req, bufio.NewWriter(&buf))
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.NotNil(t, err)
		assert.Equal(t, "404 Not Found", err.Error())
	})

	t.Run("do a request using ResponseCallback to dump the response", func(t *testing.T) {
		c, _ := New(ts.URL)
		var dump []byte
		c.ResponseCallback = func(r *http.Response) (*http.Response, error) {
			dump, _ = httputil.DumpResponse(r, true)
			return r, nil
		}
		ctx := context.Background()
		act := &message{}
		req, err := c.NewRequest(http.MethodGet, "node", act)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		var buf bytes.Buffer
		resp, err := c.Do(ctx, req, bufio.NewWriter(&buf))
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.Nil(t, err)
		assert.True(t, len(dump) > 0)
	})

	t.Run("do a request with content type unknown/unknown in response to test unmarshal behavior", func(t *testing.T) {
		c, _ := New(ts.URL)
		c.ContentType = ContentTypeText
		c.ResponseCallback = func(r *http.Response) (*http.Response, error) {
			r.Header.Set("Content-Type", "unknown/unknown")
			return r, nil
		}
		ctx := context.Background()
		req, err := c.NewRequest(http.MethodGet, "node", testMessage)
		assert.Nil(t, err)
		assert.NotNil(t, req)
		act := 42
		resp, err := c.Do(ctx, req, &act)
		if resp != nil && resp.Body != nil {
			_ = resp.Body.Close()
		}
		assert.NotNil(t, err)
	})
}


httpclient_example_test.go
package httpclient

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"os"

	"github.com/pkg/errors"
)

// drainBody reads all of b to memory and then returns two equivalent
// ReadClosers yielding the same bytes.
//
// It returns an error if the initial slurp of all bytes fails. It does not attempt
// to make the returned ReadClosers have identical error-matching behavior.
// https://golang.org/src/net/http/httputil/dump.go#L26
func drainBody(b io.ReadCloser) (r1, r2 io.ReadCloser, err error) {
	if b == http.NoBody {
		// No copying needed. Preserve the magic sentinel meaning of NoBody.
		return http.NoBody, http.NoBody, nil
	}

	var buf bytes.Buffer

	if _, err = buf.ReadFrom(b); err != nil {
		return nil, b, err
	}

	if err := b.Close(); err != nil {
		return nil, b, err
	}

	return ioutil.NopCloser(&buf), ioutil.NopCloser(bytes.NewReader(buf.Bytes())), nil
}

// This example shows how the RequestCallback function can be used to dump requests. Any pre-processing
// of requests would be possible using RequestCallback (eg. custom error checking).
func ExampleRequestCallbackFunc() {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/example" {
			http.Error(w, "invalid", http.StatusNotFound)
			return
		}
	}))
	defer ts.Close()
	c, _ := New(ts.URL)

	// Overwrite the clients standard RequestCallback function with our custom function which dumps
	// some information for each request (r) to os.Stdout and returns the unmodified request.
	// Clients functionality is not affected by this, because the standard requestCallback function of
	// the client just returns the unmodified request (r).
	c.RequestCallback = func(r *http.Request) *http.Request {
		// to convert the request in an appropriate curl command
		// command, _ := http2curl.GetCurlCommand(r)
		fmt.Fprintf(os.Stdout, "Accept: %s\nContent-Type: %s\nMethod: %s", r.Header.Get("Accept"), r.Header.Get("Content-Type"), r.Method)

		return r
	}
	req, _ := c.NewRequest(http.MethodGet, "/example", nil)

	resp, _ := c.Do(context.Background(), req, nil)
	if resp != nil && resp.Body != nil {
		_ = resp.Body.Close()
	}
	// Output:
	// Accept: application/json
	// Content-Type: application/json
	// Method: GET
}

// This example shows how the ResponseCallback function can be used to dump responses. Any post-processing
// of responses would be possible using ResponseCallback (eg. custom error checking).
func ExampleResponseCallbackFunc() {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/example" {
			http.Error(w, "invalid", http.StatusNotFound)
			return
		}

		// check post data
		body, _ := ioutil.ReadAll(r.Body)
		_, _ = w.Write(body)
	}))
	defer ts.Close()
	c, _ := New(ts.URL)

	type message struct {
		Text string
	}

	// Overwrite the clients standard ResponseCallback function with our custom function which dumps
	// some information for each response (r) to os.Stdout.
	// IMPORTANT: The clients standard responseCallback function returns an error if
	// http.StatusCode is outside of the 200 range. If you want to preserve the standard functionality consider adding the
	// if statement shown in this example.
	c.ResponseCallback = func(r *http.Response) (*http.Response, error) {
		// save the body of the response, because we consume it here but still want to return it to the caller
		var save io.ReadCloser
		save, r.Body, _ = drainBody(r.Body)
		body, _ := ioutil.ReadAll(r.Body)

		// print and then restore the body
		fmt.Fprintf(os.Stdout, "%s", body)

		r.Body = save

		// important to preserve standard clients behaviur
		if c := r.StatusCode; c >= 200 && c <= 299 {
			return r, nil
		}

		return r, errors.New(r.Status)
	}

	req, _ := c.NewRequest(http.MethodPost, "/example", &message{Text: "example"})

	resp, _ := c.Do(context.Background(), req, nil)
	if resp != nil && resp.Body != nil {
		_ = resp.Body.Close()
	}
	// Output:
	// {"Text":"example"}
}


example/jsonplaceholder/jsonplaceholder_test.go
package jsonplaceholder_test

import (
	"context"
	"testing"

	"github.com/postfinance/httpclient/example/jsonplaceholder"
)

func getClient(t *testing.T) *jsonplaceholder.Client {
	c, err := jsonplaceholder.NewClient(
		"https://jsonplaceholder.typicode.com",
	)
	if err != nil {
		t.Fatal(err)
	}

	return c
}

func TestCreate(t *testing.T) {
	c := getClient(t)
	p := jsonplaceholder.Post{
		UserID: 101,
		Title:  "Blog post title",
		Body:   "Blog post body",
	}

	post, _, err := c.Post.Create(context.Background(), &p)
	if err != nil {
		t.Error(err)
	}

	if post.ID == 0 {
		t.Error("post ID cannot be zero")
	}

	t.Logf("%#v\n", post)
}

func TestGet(t *testing.T) {
	c := getClient(t)

	post, _, err := c.Post.Get(context.Background(), 1)
	if err != nil {
		t.Error(err)
	}

	t.Logf("%#v\n", post)
}

func TestList(t *testing.T) {
	c := getClient(t)

	posts, _, err := c.Post.List(context.Background())
	if err != nil {
		t.Error(err)
	}

	if len(posts) == 0 {
		t.Error("no posts found")
	}

	t.Logf("number of posts: %d\n", len(posts))
}

func TestDelete(t *testing.T) {
	c := getClient(t)

	_, err := c.Post.Delete(context.Background(), 1)
	if err != nil {
		t.Error(err)
	}
}


example/jsonplaceholder/jsonplaceholder.go
// Package jsonplaceholder implements an example httpclient
package jsonplaceholder

import (
	"context"
	"fmt"
	"net/http"

	"github.com/postfinance/httpclient"
)

// Post a post entry
type Post struct {
	UserID int    `json:"userId"`
	ID     int    `json:"id"`
	Title  string `json:"title"`
	Body   string `json:"body"`
}

// PostService interface defines service methods
type PostService interface {
	Get(context.Context, int) (*Post, *http.Response, error)
	List(context.Context) ([]Post, *http.Response, error)
	Create(context.Context, *Post) (*Post, *http.Response, error)
	Delete(context.Context, int) (*http.Response, error)
}

// PostImpl implements the PostService interface
type PostImpl struct {
	client *httpclient.Client
}

var _ PostService = &PostImpl{}

// Get a post by ID
func (p *PostImpl) Get(ctx context.Context, id int) (*Post, *http.Response, error) {
	req, err := p.client.NewRequest(http.MethodGet, fmt.Sprintf("/posts/%d", id), nil)
	if err != nil {
		return nil, nil, err
	}

	post := Post{}

	resp, err := p.client.Do(ctx, req, &post)
	if err != nil {
		return nil, nil, err
	}

	return &post, resp, nil
}

// List all posts
func (p *PostImpl) List(ctx context.Context) ([]Post, *http.Response, error) {
	req, err := p.client.NewRequest(http.MethodGet, "/posts", nil)
	if err != nil {
		return nil, nil, err
	}

	posts := []Post{}

	resp, err := p.client.Do(ctx, req, &posts)
	if err != nil {
		return nil, nil, err
	}

	return posts, resp, nil
}

// Create a new post
func (p *PostImpl) Create(ctx context.Context, newPost *Post) (*Post, *http.Response, error) {
	req, err := p.client.NewRequest(http.MethodPost, "/posts", newPost)
	if err != nil {
		return nil, nil, err
	}

	post := Post{}

	resp, err := p.client.Do(ctx, req, &post)
	if err != nil {
		return nil, nil, err
	}

	return &post, resp, nil
}

// Delete a post by ID
func (p *PostImpl) Delete(ctx context.Context, id int) (*http.Response, error) {
	req, err := p.client.NewRequest(http.MethodDelete, fmt.Sprintf("/posts/%d", id), nil)
	if err != nil {
		return nil, err
	}

	resp, err := p.client.Do(ctx, req, nil)
	if err != nil {
		return nil, err
	}

	return resp, nil
}


httpclient.go
// Package httpclient provides the basic infrastructure for doing RESTish http requests
// an application specific client will be generated with the client-gen-go tool.
package httpclient

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"reflect"
	"time"

	"golang.org/x/time/rate"

	yaml "gopkg.in/yaml.v2"

	"github.com/google/go-querystring/query"
	"github.com/pkg/errors"
)

// Constants
const (
	ContentTypeText = "text/plain"
	ContentTypeJSON = "application/json"
	ContentTypeYAML = "application/yaml"
)

// Variables
var (
	ErrUnknownContentType = errors.New("unknown media type")
	ErrTooManyRequest     = errors.New("too many requests")
)

// Client provides ....
type Client struct {
	// HTTP client used to communicate with the server
	client *http.Client

	// rate limiter
	limiter *rate.Limiter

	// Base URL for API requests.
	BaseURL *url.URL

	// ContentType is used as Content-Type and Accept in request headers.
	ContentType string

	// username/password for basic authentication
	username string
	password string

	// custom http header(s)
	header http.Header

	Marshaler   MarshalerFunc
	Unmarshaler UnmarshalerFunc

	RequestCallback  RequestCallbackFunc
	ResponseCallback ResponseCallbackFunc
}

// Opt are options for New.
type Opt func(*Client) error

// RequestCallbackFunc for custom pre-processing of requests
// possible use cases: custom error checking, dumping requests for debugging etc.
type RequestCallbackFunc func(*http.Request) *http.Request

// ResponseCallbackFunc for custom post-processing of responses
// possible use cases: custom error checking, dumping responses for debugging etc.
type ResponseCallbackFunc func(*http.Response) (*http.Response, error)

// MarshalerFunc for custom marshaling function
type MarshalerFunc func(io.Writer, interface{}, string) (string, error)

// UnmarshalerFunc for custom unmarshaling function
type UnmarshalerFunc func(io.Reader, interface{}, string) error

// QueryOptions adds query options opt to URL u
// opt has to be a struct tagged according to https://github.com/google/go-querystring
// e.g.:
// type options struct {
//     Page    int    `url:"page,omitempty"`
//     PerPage int    `url:"per_page,omitempty"`
//     Search  string `url:"search,omitempty"`
// }
// opt := options{1, 10, "name=testHost"}
// ... will be added to URL u as "?page=1&per_page=10&search=name%3DtestHost"
func QueryOptions(u string, opt interface{}) (string, error) {
	v := reflect.ValueOf(opt)

	if v.Kind() == reflect.Ptr && v.IsNil() {
		return u, nil
	}

	origURL, err := url.Parse(u)
	if err != nil {
		return u, err
	}

	origValues := origURL.Query()

	newValues, err := query.Values(opt)
	if err != nil {
		return u, err
	}

	for k, v := range newValues {
		origValues[k] = v
	}

	origURL.RawQuery = origValues.Encode()

	return origURL.String(), nil
}

// New returns a new client instance.
func New(baseURL string, opts ...Opt) (*Client, error) {
	u, err := url.Parse(baseURL)
	if err != nil {
		return nil, err
	}

	c := &Client{
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
		BaseURL:          u,
		ContentType:      ContentTypeJSON,
		Marshaler:        marshal,
		Unmarshaler:      unmarshal,
		RequestCallback:  requestCallback,
		ResponseCallback: responseCallback,
	}

	for _, opt := range opts {
		if err := opt(c); err != nil {
			return nil, err
		}
	}

	return c, nil
}

// WithPassword is a client option for setting the password for basic authentication.
func WithPassword(p string) Opt {
	return func(c *Client) error {
		if p == "" {
			return errors.New("password cannot be empty")
		}

		c.password = p

		return nil
	}
}

// WithUsername is a client option for setting the username.
func WithUsername(u string) Opt {
	return func(c *Client) error {
		if u == "" {
			return errors.New("username cannot be empty")
		}

		c.username = u

		return nil
	}
}

// WithHTTPClient is a client option for setting another http client than the default one
func WithHTTPClient(c *http.Client) Opt {
	return func(cli *Client) error {
		cli.client = c
		return nil
	}
}

// WithRateLimiter see https://godoc.org/golang.org/x/time/rate
func WithRateLimiter(l *rate.Limiter) Opt {
	return func(cli *Client) error {
		cli.limiter = l
		return nil
	}
}

// WithContentType is a client option for setting the content type
func WithContentType(ct string) Opt {
	return func(c *Client) error {
		if ct == "" {
			return errors.New("content type cannot be empty")
		}

		c.ContentType = ct

		return nil
	}
}

// WithHeader is a client option for setting custom http header(s) for each request
// Content-Type and Accept headers will be appended by the clients ContentType setting
// Authorization header is overwritten if WithUsername/WithPassowrd was used to setup the client
func WithHeader(header http.Header) Opt {
	return func(c *Client) error {
		c.header = header
		return nil
	}
}

// NewRequest creates an API request. A relative URL can be provided in urlStr, which will be resolved to the
// BaseURL of the Client. Relative URLs should always be specified without a preceding slash. If specified, the
// value pointed to by body will be encoded and included in as the request body.
func (c *Client) NewRequest(method, urlStr string, body interface{}) (*http.Request, error) {
	rel, err := url.Parse(urlStr)
	if err != nil {
		return nil, err
	}

	u := c.BaseURL.ResolveReference(rel)

	if c.Marshaler == nil {
		panic("Marshaler is nil")
	}

	buf := new(bytes.Buffer)

	contentType, err := c.Marshaler(buf, body, c.ContentType)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest(method, u.String(), buf)
	if err != nil {
		return nil, err
	}

	if c.header != nil {
		req.Header = c.header.Clone()
	}

	if len(c.username) > 0 && len(c.password) > 0 {
		req.SetBasicAuth(c.username, c.password)
	}

	req.Header.Add("Content-Type", contentType)
	req.Header.Add("Accept", contentType)

	if c.RequestCallback == nil {
		panic("RequestCallback is nil")
	}

	return c.RequestCallback(req), nil
}

// marshal is the default marshaler
func marshal(w io.Writer, v interface{}, mediaType string) (string, error) {
	if v == nil {
		return mediaType, nil
	}

	switch mediaType {
	case ContentTypeJSON:
		return ContentTypeJSON, MarshalJSON(w, v, mediaType)
	case ContentTypeYAML:
		return ContentTypeYAML, MarshalYAML(w, v, mediaType)
	case ContentTypeText:
		_, err := fmt.Fprint(w, v)
		return ContentTypeText, err
	default:
		return mediaType, errors.Wrap(ErrUnknownContentType, mediaType)
	}
}

// MarshalJSON marshal JSON
func MarshalJSON(w io.Writer, v interface{}, mediaType string) error {
	return json.NewEncoder(w).Encode(v)
}

// MarshalYAML marshal JSON
func MarshalYAML(w io.Writer, v interface{}, mediaType string) error {
	b, err := yaml.Marshal(v)
	if err != nil {
		return err
	}

	_, err = w.Write(b)

	return err
}

// Do sends an API request and returns the API response. The API response will be decoded and stored in the value
// pointed to by v, or returned as an error if an API error has occurred. If v implements the io.Writer interface,
// the raw response will be written to v, without attempting to decode it.
func (c *Client) Do(ctx context.Context, req *http.Request, v interface{}) (*http.Response, error) {
	// rate limit
	if c.limiter != nil {
		if err := c.limiter.Wait(ctx); err != nil {
			return nil, ErrTooManyRequest
		}
	}

	resp, err := c.client.Do(req.WithContext(ctx))
	if err != nil {
		return resp, err
	}

	defer func() {
		if rerr := resp.Body.Close(); rerr == nil {
			err = rerr
		}
	}()

	if c.ResponseCallback == nil {
		panic("ResponseCallback is nil")
	}

	resp, err = c.ResponseCallback(resp)
	if err != nil {
		return resp, err
	}

	if c.Unmarshaler == nil {
		panic("Unmarshaler is nil")
	}

	err = c.Unmarshaler(resp.Body, v, c.ContentType)

	return resp, err
}

// unmarshal is the default unmarshaler
func unmarshal(r io.Reader, v interface{}, mediaType string) error {
	if v == nil {
		return nil
	}
	// if v is a io.Writer copy the request body to v
	if w, ok := v.(io.Writer); ok {
		_, err := io.Copy(w, r)
		return err
	}

	switch mediaType {
	case ContentTypeJSON:
		return UnmarshalJSON(r, v, mediaType)
	case ContentTypeYAML:
		return UnmarshalYAML(r, v, mediaType)
	case ContentTypeText:
		if x, ok := v.(*string); ok {
			buf := new(bytes.Buffer)
			if _, err := buf.ReadFrom(r); err != nil {
				return errors.Wrap(err, "read into buffer")
			}

			*x = buf.String()

			return nil
		}

		return errors.New("target type is not string")
	default:
		return errors.Wrap(ErrUnknownContentType, mediaType)
	}
}

// UnmarshalJSON unmarshal JSON
func UnmarshalJSON(r io.Reader, v interface{}, mediaType string) error {
	return json.NewDecoder(r).Decode(v)
}

// UnmarshalYAML unmarshal YAML
func UnmarshalYAML(r io.Reader, v interface{}, mediaType string) error {
	data, err := ioutil.ReadAll(r)
	if err != nil {
		return err
	}

	return yaml.Unmarshal(data, v)
}

// requestCallback returns the unmodified request
func requestCallback(r *http.Request) *http.Request {
	return r
}

// responseCallback checks the API response for errors, and returns them if present. A response is considered an
// error if it has a status code outside the 200 range. API error responses are expected to have no response body.
func responseCallback(r *http.Response) (*http.Response, error) {
	if c := r.StatusCode; c >= 200 && c <= 299 {
		return r, nil
	}

	return r, errors.New(r.Status)
}
